### ISSUE TITLE
Lint to check unnecessary trait bounds

### ISSUE THREAD

**Issue by laralove143** [rust-lang/rust-clippy#10828]
### What it does

Some trait bounds aren't required, though they may be added for semantic purposes, this is rare. An example is say we have this function:
```rust
async fn<T: Send + Sync>foo() { ... }
```
then it no longer needs async so we change it to
```rust
fn<T: Send + Sync>foo() { ... }
```
but we forgot to remove the `Send + Sync` bounds and they slip unnoticed, this is bad because users can no longer use types that don't implement `Send + Sync` even though they actually can

### Lint Name

unnecessary_trait_bounds

### Category

correctness

### Advantage

- Removes unnecessary restrictions
- Makes code more clear
- Possibly even frees a dependency if the trait is from a crate and it's not used otherwsie

### Drawbacks

Basically warns on almost every trait bound in struct definitions:
```rust
struct<T: Send> Foo<T> { ... }
```
Whether this is good design is controversial, this could be separated to another lint but I don't see it necessary as users can simply allow it

### Example

```rust
fn print<T: Debug + Display>(x: T) {
  println!("{x:?}");
}
```

Could be written as:

```rust
fn print<T: Debug>(x: T) {
  println!("{x:?}");
}
```

**Comment by jhpratt** [comment]
> Basically warns on almost every trait bound in struct definitions:
>
> ```rust
> struct<T: Send> Foo<T> { ... }
> ```
>
> Whether this is good design is controversial

See [this StackOverflow post](https://stackoverflow.com/questions/49229332/should-trait-bounds-be-duplicated-in-struct-and-impl). This pattern is followed in the standard library on types such as `Cell` and `HashMap`.
